### ISSUE TITLE
Allow kernel objects to specify a "cleanup" handler to be executed on behalf of pended threads that timeout

### ISSUE THREAD

**Issue by peter-mitsis** [zephyrproject-rtos/zephyr#45033]
**Is your enhancement proposal related to a problem? Please describe.**
Some threads, after they time-out from waiting on a kernel object require some cleanup work to be done on the kernel object. Rather than being done immediately, this is typically only be done once that thread has been scheduled to execute.  This may result in suboptimal behavior or system performance until that is done.

One example of this is a mutex that is involved in priority inheritance. In such a scenario, a thread that had its priority bumped due to a pending higher priority thread will have its priority running at a higher priority longer than necessary or desired. It may continue to do so until the timed-out thread finally executes to correct the issue.

**Describe the solution you'd like**
I would like to see something along the lines of a kernel object being able to specify a cleanup handler when it places a thread on a wait queue that is to be executed only if the the pended thread times out.

**Describe alternatives you've considered**
One alternative is to perform a check at the start of an operation on the affected kernel object to see if there is any cleanup work to do and do it BEFORE doing anything else. However, this I think adds unnecessary complexity.

**Comment by zephyrbot** [comment]
Hi @peter-mitsis, @andyross,

This issue, marked as an Enhancement, was opened a while ago and did not get any traction. It was just assigned to you based on the labels. If you don't consider yourself the right person to address this issue, please re-assing it to the right person.

Please take a moment to review if the issue is still relevant to the project.  If it is, please provide feedback and direction on how to move forward. If it is not, has already been addressed, is a duplicate, or is no longer relevant, please close it with a short comment explaining the reason.


Thanks!
